kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java

```diff
@@ -0,0 +1,211 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.configuration.kafka.streams.health;
+
+import io.micronaut.configuration.kafka.streams.ConfiguredStreamBuilder;
+import io.micronaut.configuration.kafka.streams.KafkaStreamsConfiguration;
+import io.micronaut.configuration.kafka.streams.KafkaStreamsFactory;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.health.HealthStatus;
+import io.micronaut.management.health.aggregator.HealthAggregator;
+import io.micronaut.management.health.indicator.HealthIndicator;
+import io.micronaut.management.health.indicator.HealthResult;
+import io.reactivex.Flowable;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.processor.TaskMetadata;
+import org.apache.kafka.streams.processor.ThreadMetadata;
+import org.reactivestreams.Publisher;
+
+import javax.inject.Singleton;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link HealthIndicator} for Kafka Streams.
+ *
+ * @author Christian Oestreich
+ * @since 2.0.1
+ */
+@Singleton
+@Requires(classes = HealthIndicator.class)
+@Requires(property = KafkaStreamsConfiguration.PREFIX + ".health.enabled", value = "true", defaultValue = "true")
+public class KafkaStreamsHealth implements HealthIndicator {
+
+    private static final String NAME = "kafkaStreams";
+
+    private final KafkaStreamsFactory kafkaStreamsFactory;
+    private final HealthAggregator<?> healthAggregator;
+
+    /**
+     * Constructor for the health check.
+     *
+     * @param kafkaStreamsFactory The stream factory to get streams from
+     * @param healthAggregator    Health aggregator
+     */
+    public KafkaStreamsHealth(final KafkaStreamsFactory kafkaStreamsFactory, final HealthAggregator<?> healthAggregator) {
+        this.kafkaStreamsFactory = kafkaStreamsFactory;
+        this.healthAggregator = healthAggregator;
+    }
+
+    /**
+     * Get the health result of the streams.  Will attempt to interrogate
+     * details of each stream as well.  The application.id will be used
+     * for each configured stream as the primary health check name.
+     *
+     * @return Health Result Aggregate
+     */
+    @Override
+    public Publisher<HealthResult> getResult() {
+        Flowable<HealthResult> kafkaStreamHealth = Flowable.fromIterable(kafkaStreamsFactory.getStreams().keySet())
+                .map(kafkaStreams -> {
+                    String applicationId = getApplicationId(kafkaStreams);
+                    try {
```

Comment:
Perhaps use `filter`, `defaultIfEmpty` and `onErrorReturn` instead?

Reply:
yeah I am not the rx guru like you. `maybe(that would be better?).defaultIfEmpty("@bwyka does code change").onErrorReturn("do not change the code at all?")`

Reply:
At your command:

```java
    @Override
    public Publisher<HealthResult> getResult() {
        Flowable<HealthResult> kafkaStreamHealth = Flowable.fromIterable(Collections.singleton(new KafkaStreams((Topology) null, (Properties) null)))
                .map(kafkaStreams -> Pair.of(getApplicationId(kafkaStreams), kafkaStreams))
                .flatMap(pair -> Flowable.just(pair)
                        .filter(p -> p.getValue().state().isRunning())
                        .map(p -> HealthResult.builder(p.getKey(), HealthStatus.UP)
                                .details(buildDetails(p.getValue())))
                        .defaultIfEmpty(HealthResult.builder(pair.getKey(), HealthStatus.DOWN)
                                .details(Collections.singletonMap("error", "The processor is down")))
                        .onErrorReturn(e -> HealthResult.builder(pair.getKey(), HealthStatus.DOWN)
                                .details(buildDownDetails(e, pair.getValue().state()))))
                .map(HealthResult.Builder::build);
        return healthAggregator.aggregate(NAME, kafkaStreamHealth);
    }
```